### ISSUE TITLE
request-server and long directory list

### ISSUE THREAD

**Issue by purha** [pkg/sftp#184]
Hi,

I get errors when listing directory with high amount of files:
```
sftp> cd bigdir
sftp> ls
Received message too long 4317039
```

I used simple bash script to create test files:
```
#!/bin/bash
NUM=25000

for i in $(seq 1 $NUM); do
  echo "this is file number $i" > "testfile_xxxxxxxxx-xxxxxxxx-xxxxxxxxx-xxxxxxxx-xxxxxxxx-xxxxxxx-$i.txt" 
done
```

I manually tried to adjust maxTxPacket but with no luck.

Br,
Mikko

**Comment by eikenb** [comment]
Is this when running examples/sftp-server and connecting with openssh's sftp client?
If not, please add some details about what you are running so I can reproduce the problem.

Thanks.

**Comment by eikenb** [comment]
Using the openssh client and examples/sftp-server on Linux, I wasn't able to reproduce the issue.

**Comment by purha** [comment]
I created test case for it based in InMemoryHandler:

```
package main

// This serves as an example of how to implement the request server handler as
// well as a dummy backend for testing. It implements an in-memory backend that
// works as a very simple filesystem with simple flat key-value lookup system.

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sync"
	"syscall"
	"time"

	"github.com/pkg/sftp"
)

func fakeFileInfoSys() interface{} {
	return syscall.Stat_t{
		Uid:  65534,
		Gid:  65534,
		Size: 65534,
	}
}

// InMemTestHandler returns a Hanlders object with the test handlers
func InMemTestHandler() sftp.Handlers {
	root := &root{
		files: make(map[string]*memFile),
	}
	root.memFile = newMemFile("/", true)

	for i := 0; i < 250000; i++ {
		fn := fmt.Sprintf("/testfile_xxxxxxxxx-xxxxxxxx-xxxxxxxxx-xxxxxxxx-xxxxxxxx-xxxxxxx-%v.txt", i)
		content := fmt.Sprintf("this is file number %v", i)
		root.files[fn] = newMemFileExample(fn, content)
	}
	return sftp.Handlers{root, root, root, root}
}

// Handlers
func (fs *root) Fileread(r sftp.Request) (io.ReaderAt, error) {
	fs.filesLock.Lock()
	defer fs.filesLock.Unlock()
	file, err := fs.fetch(r.Filepath)
	if err != nil {
		return nil, err
	}
	if file.symlink != "" {
		file, err = fs.fetch(file.symlink)
		if err != nil {
			return nil, err
		}
	}
	return file.ReaderAt()
}

func (fs *root) Filewrite(r sftp.Request) (io.WriterAt, error) {
	fs.filesLock.Lock()
	defer fs.filesLock.Unlock()
	file, err := fs.fetch(r.Filepath)
	if err == os.ErrNotExist {
		dir, err := fs.fetch(filepath.Dir(r.Filepath))
		if err != nil {
			return nil, err
		}
		if !dir.isdir {
			return nil, os.ErrInvalid
		}
		file = newMemFile(r.Filepath, false)
		fs.files[r.Filepath] = file
	}
	return file.WriterAt()
}

func (fs *root) Filecmd(r sftp.Request) error {
	fs.filesLock.Lock()
	defer fs.filesLock.Unlock()
	switch r.Method {
	case "Setstat":
		return nil
	case "Rename":
		file, err := fs.fetch(r.Filepath)
		if err != nil {
			return err
		}
		if _, ok := fs.files[r.Target]; ok {
			return &os.LinkError{Op: "rename", Old: r.Filepath, New: r.Target,
				Err: fmt.Errorf("dest file exists")}
		}
		fs.files[r.Target] = file
		delete(fs.files, r.Filepath)
	case "Rmdir", "Remove":
		_, err := fs.fetch(filepath.Dir(r.Filepath))
		if err != nil {
			return err
		}
		delete(fs.files, r.Filepath)
	case "Mkdir":
		_, err := fs.fetch(filepath.Dir(r.Filepath))
		if err != nil {
			return err
		}
		fs.files[r.Filepath] = newMemFile(r.Filepath, true)
	case "Symlink":
		_, err := fs.fetch(r.Filepath)
		if err != nil {
			return err
		}
		link := newMemFile(r.Target, false)
		link.symlink = r.Filepath
		fs.files[r.Target] = link
	}
	return nil
}

func (fs *root) Fileinfo(r sftp.Request) ([]os.FileInfo, error) {
	fs.filesLock.Lock()
	defer fs.filesLock.Unlock()
	switch r.Method {
	case "List":
		list := []os.FileInfo{}
		for fn, fi := range fs.files {
			if filepath.Dir(fn) == r.Filepath {
				list = append(list, fi)
			}
		}
		return list, nil
	case "Stat":
		file, err := fs.fetch(r.Filepath)
		if err != nil {
			return nil, err
		}
		return []os.FileInfo{file}, nil
	case "Readlink":
		file, err := fs.fetch(r.Filepath)
		if err != nil {
			return nil, err
		}
		if file.symlink != "" {
			file, err = fs.fetch(file.symlink)
			if err != nil {
				return nil, err
			}
		}
		return []os.FileInfo{file}, nil
	}
	return nil, nil
}

// In memory file-system-y thing that the Hanlders live on
type root struct {
	*memFile
	files     map[string]*memFile
	filesLock sync.Mutex
}

func (fs *root) fetch(path string) (*memFile, error) {
	if path == "/" {
		return fs.memFile, nil
	}
	if file, ok := fs.files[path]; ok {
		return file, nil
	}
	return nil, os.ErrNotExist
}

// Implements os.FileInfo, Reader and Writer interfaces.
// These are the 3 interfaces necessary for the Handlers.
type memFile struct {
	name        string
	modtime     time.Time
	symlink     string
	isdir       bool
	content     []byte
	contentLock sync.RWMutex
}

// factory to make sure modtime is set
func newMemFile(name string, isdir bool) *memFile {
	return &memFile{
		name:    name,
		modtime: time.Now(),
		isdir:   isdir,
	}
}

// factory to make sure modtime is set
func newMemFileExample(name string, content string) *memFile {
	return &memFile{
		name:    name,
		modtime: time.Now(),
		isdir:   false,
		content: []byte(content),
	}
}

// Have memFile fulfill os.FileInfo interface
func (f *memFile) Name() string { return filepath.Base(f.name) }
func (f *memFile) Size() int64  { return int64(len(f.content)) }
func (f *memFile) Mode() os.FileMode {
	ret := os.FileMode(0644)
	if f.isdir {
		ret = os.FileMode(0755) | os.ModeDir
	}
	if f.symlink != "" {
		ret = os.FileMode(0777) | os.ModeSymlink
	}
	return ret
}
func (f *memFile) ModTime() time.Time { return f.modtime }
func (f *memFile) IsDir() bool        { return f.isdir }
func (f *memFile) Sys() interface{} {
	return fakeFileInfoSys()
}

// Read/Write
func (f *memFile) ReaderAt() (io.ReaderAt, error) {
	if f.isdir {
		return nil, os.ErrInvalid
	}
	return bytes.NewReader(f.content), nil
}

func (f *memFile) WriterAt() (io.WriterAt, error) {
	if f.isdir {
		return nil, os.ErrInvalid
	}
	return f, nil
}
func (f *memFile) WriteAt(p []byte, off int64) (int, error) {
	// fmt.Println(string(p), off)
	// mimic write delays, should be optional
	time.Sleep(time.Microsecond * time.Duration(len(p)))
	f.contentLock.Lock()
	defer f.contentLock.Unlock()
	plen := len(p) + int(off)
	if plen >= len(f.content) {
		nc := make([]byte, plen)
		copy(nc, f.content)
		f.content = nc
	}
	copy(f.content[off:], p)
	return len(p), nil
}

```

**Comment by purha** [comment]
main program is same as examples/request-server/main.go where InMemHandler is replaced with one from this InMemTestHandler

**Comment by eikenb** [comment]
Thanks for the test case. I was able to reproduce it using that. I'll look into it.

**Comment by eikenb** [comment]
Figured out the issue. The protocol doesn't specify a maximum number of files returned for readdir, but it does define things so that it make sense to return the files in batches in section 6.7 of the RFC [1]. Clients seem to enforce their own limits on this so batching is needed. The openssh-server uses batches of 100.

[1] https://tools.ietf.org/html/draft-ietf-secsh-filexfer-02#section-6.7
TLDR; the client keeps calling readdir until the server replies with an EOF error.

Currently you can't batch up the results in the Fileinfo() handler because the wrapping call returns the EOF after 1 batch is returned (ie. it doesn't support batching). The wrapping call here is the fileinfo() function in request.go.

Options I've come up with at this point. In order of the amount of work pushed to the handler implementer.
1) The Fileinfo() wrapper could batch things automatically behind the scenes.
2) Have the Fileinfo() return batches and check for an empty Fileinfo slice, and set EOF in that case. 
3) Have the Fileinfo() return batches and send the EOF when done.

If you like one of those or have an idea of your own please chime in.

**Comment by eikenb** [comment]
In order to allow the handler writer to work in batches the sftp.Request object would need a place to write session data. Probably ye ole map[string]interface{}.

**Comment by eikenb** [comment]
I'm starting to work on this. I've decided options 1. is out as that would require storing all the file info in memory as it doles it out to the client. Given that the filelist could feasibly be of any size, this seems like a bad idea. So I'm doing 3 and maybe 2 as well as it seems like an easy add on.

**Comment by eikenb** [comment]
@purha Quick question. For the request backend you are working on, what sort of data type would be best to keep track of your place in the file list for batching. In my S3 backend, AWS gives me a token (string) and just having it as a string seems general purpose enough. The other option would be to make it an interface{}, which means type assertions every time you access it and I would like to avoid that if possible.

**Comment by purha** [comment]
@eikenb I'm also using S3 (and minio) backend, so the string is good option for me.

**Comment by eikenb** [comment]
@purha Let me know if you have any questions with how I addressed the issue. The commit message goes into some details and I updated the request-example and tests with batching.

**Comment by purha** [comment]
@eikenb thank you! I'll take a look.

**Comment by eikenb** [comment]
@purha Sorry to resurrect this ticket, but I've been thinking about my solution and find the requirement of the handler having to deal with LsNext/LsSave to be annoying and had was thinking about an alternative implementation. I'm curious what you think.

The idea is to try to mimic the Fileread/Filewrite handlers more. To do this I'd create a new interface..

    FilelisterAt([]os.Fileinfo, int64) (int, error)

Implementation/doc wise it would look like..

    FilelisterAt(ls []os.Fileinfo, offset int) (n int, err error)

This would work like ReaderAt() where it would list entries of len([]os.Fileinfo) starting at offset. It would return the number of entries listed and an error. If n < len([]os.Fileinfo) it would return a non-nill error, EOF when it hit the end of the list.

So you'd need to return something that met that interface and the wrapper would take care of repeatedly calling it, like the Fileread/Filewrite hander methods work.

I'm not 100% sure, but it does seem that something like this would be better.

**Comment by purha** [comment]
@eikenb this is also perfectly fine solution for me, and it's bit cleaner too.

**Comment by eikenb** [comment]
I just pushed branch "issue184" that contains the new version that implements what I described above and IMO it is definitely much cleaner. Though it makes the file listing handler a bit different. I'm also not 100% happy with the request-example.go implementation of it and I want to clean that up to be a good example of how it might be implemented. 

Let me know if you have any thoughts on it. Thanks.

**Comment by eikenb** [comment]
@purha Just pushed all the changes I was thinking about for this to the issue184 branch. IMO it is ready for merging.

**Comment by eikenb** [comment]
I've been working with the new interface based setup and am not sure that using an interface is the best approach. Seems like just returning a function closure would be much easier and a bit more flexible.

That is changing from Filinfo() returning an interface like this...

```
type FileInfoer interface {
	Fileinfo(Request) (FileInfoAter, error)
}
type FileInfoAter interface {
	FileInfoAt([]os.FileInfo, int64) (int, error)
}
```

To something more like this where is returns a function...

```
type FileInfoer interface {
	Fileinfo(Request) (FileInfoAt, error)
}
type FileInfoAt func([]os.FileInfo, int64) (int, error)
```

The tradeoffs seem to be that the former is closer in spirit to FileReader and FileWriter, thus more uniform as an API. While the latter is simpler to use as you can just close over the relevant context w/ a returned anonymouse function vs. adding a struct w/ the method and copying the request and other relevant data into it.

I've already done the former. I think I'm going to go try the latter in a separate branch so I can better compare them. Anyone who cares please feel free to chime in.

**Comment by puellanivis** [comment]
I think the thing that strikes me the most is the names seem so unwieldy… in some ways, they seem inconsistent with the way Go does file management, (as a FileInfoXY() method is using a noun, not a verb.)

Looking at how the API would work, both seem like good choices, just… names? I dunno… I like the design, just it has my lizard brain making me uncomfortable for some reason… I’m pretty sure it’s the names though.

**Comment by eikenb** [comment]
@puellanivis Thanks for the input and yeah, the names are bit clumsy. I was trying to stick with the pattern set by the io package naming for ReaderAT/ReadAt and tried a few different names and never found a great one.

We have 1 or 2 interfaces and 2 methods to name (depending on which way we go).

Currently we have the not-so-pretty FileInfoer interface with the Fileinfo() method. It returns the FileInfoAter interface with the FileInfoAt() method. The only idea off the top of my head as an alternative would be to switch to Filelist. I was using Fileinfo as the interface is used for both lists  and stat, but it is always a slice so Filelist can work.

That would make the names; Filelister interface, Filelist() method and FilelisterAt interface, FilelistAt() method. Those are slightly less unwieldy.

What do you think? Please suggest an alternative name if you have an idea.

**Comment by puellanivis** [comment]
From my file/blob-abstraction library, I’ve used `List`, which parallels with the *nix `ls` command.

**Comment by eikenb** [comment]
So maybe just Lister/List() and ListerAt/ListAt()?

@purha Any preferences?

**Comment by eikenb** [comment]
No. I want to keep the naming scheme with the handlers. So I want to keep Filelister/Filelist() as the handler interface. But for the returned interface or function I think the type/name would be better to follow your example. So that woul be ListerAt/LIstAt(). This also lines up with the file reading handler naming scheme of Filereader/Fileread()/ReaderAt/ReadAt().

I think that is the best compromise between more readable naming and sticking with the scheme.

Now I just need to decided if I like the interface or function return better.

**Comment by eikenb** [comment]
Spent the weekend porting my s3 backend to each version (function/interface) of the new API and it doesn't make that much difference in practice. So I'm going to go with the interface based API because it is more consistent with the Reader/Writer API.

So I'm going to review that branch (issue184) to be sure naming is reasonably consistent then merge it.
